### PR TITLE
Update to Jest 29

### PR DESCRIPTION
**Pull request by geriux**
This PR updates the Jest version to `29` to match Gutenberg's current version introduced [in this PR.](https://github.com/WordPress/gutenberg/pull/47388)

To test CI checks should pass.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -34,7 +34,9 @@ module.exports = {
 		'<rootDir>/jetpack/',
 		'/__device-tests__/',
 	],
-	testURL: 'http://localhost/',
+	testEnvironmentOptions: {
```
Comment (geriux) [inline jest.config.js:37]: `testURL` is deprecated so we needed to update it to the new format.

File: jest_setup.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Jetpack mocks
+global.window.JP_CONNECTION_INITIAL_STATE = {};
```
Comment (geriux) [inline jest_setup.js]: This was added to remove the console error: 

`Jetpack Connection package: Initial state is missing. `

This was happening before the Jest upgrade.

File: jest_setup.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+jest.mock( './jetpack/projects/js-packages/config/src', () => ( {
```
Comment (geriux) [inline jest_setup.js]: Added to remove the console error:

`jetpackConfig is missing in your webpack config file. See @**********/jetpack-config`

File: jest_setup.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+	__esModule: true,
+} ) );
+
+jest.mock(
```
Comment (geriux) [inline jest_setup.js]: After the Jest 29 update, there is an issue where it can't find the native export of this file so I added a mock.

**Comment by geriux** [comment]
FYI there are a few CI failures for this PR due to other reasons like `node` failing to download.

File: package.json [hunk 5]
```diff
@@ -39,8 +39,9 @@
 		"gettext-parser": "^4.0.3",
 		"husky": "^3.0.2",
 		"identity-obj-proxy": "^3.0.0",
-		"jest": "^27.4.5",
+		"jest": "29.5.0",
 		"jest-image-snapshot": "^6.1.0",
+		"jest-jasmine2": "29.5.0",
```
Comment (geriux) [inline package.json:44]: `jest-jasmine2` is no longer included with Jest and it has [to be installed separately](https://jestjs.io/docs/28.x/upgrading-to-jest28#test-runner).

File: jest.config.js [hunk 6]
```diff
@@ -25,7 +25,7 @@ module.exports = {
 	// Automatically clear mock calls and instances between every test
 	clearMocks: true,
 	preset: './gutenberg/node_modules/react-native/jest-preset.js',
-	setupFiles: [ '<rootDir>/' + configPath + '/setup.js' ],
+	setupFiles: [ '<rootDir>/' + configPath + '/setup.js', './jest_setup.js' ],
```
Comment (fluiddot) [inline jest.config.js]: Nitpick: I'm wondering if we could use a similar name format for `jest_setup.js` as this file, i.e. `jest.setup.js`, WDYT?

File: jest_setup.js (hunk 2 above)
Comment (fluiddot) [reply jest_setup.js]: I saw this approach is also used in the Jetpack unit tests: https://github.com/Automattic/jetpack/blob/trunk/projects/js-packages/connection/jest.setup.js

File: jest_setup.js (hunk 3 above)
Comment (fluiddot) [reply jest_setup.js]: I checked how this is mocked in the Jetpack unit tests and found [this configuration](https://github.com/Automattic/jetpack/blob/2a306ced82a35a6a4bf199edc8af77494584b53d/tools/js-tools/jest/config.base.js#L27-L29). I tested locally and seems to work by adding the following to `jest.config.js`:
```patch
diff --git forkSrcPrefix/jest.config.js forkDstPrefix/jest.config.js
index 81bbacc988d81e281a87e34086a20bf93436e847..4846110d6ae9b9f0b206d44fc8510a1c3f24bf55 100644
--- forkSrcPrefix/jest.config.js
+++ forkDstPrefix/jest.config.js
@@ -51,6 +51,8 @@ module.exports = {
 			'|'
 		) })$` ]: '<rootDir>/gutenberg/packages/$1/src',
 		'test/helpers$': '<rootDir>/' + configPath + '/helpers.js',
+		jetpackConfig:
+			'<rootDir>/jetpack/tools/js-tools/jest/jest-jetpack-config.js',
 	},
 	haste: {
 		defaultPlatform: rnPlatform,

```

File: jest.config.js (hunk 6 above)
Comment (geriux) [reply jest.config.js]: Sure! I'll update it 👍

File: jest_setup.js (hunk 3 above)
Comment (geriux) [reply jest_setup.js]: Nice suggestion! I'll apply this change.

File: jest_setup.js (hunk 2 above)
Comment (geriux) [reply jest_setup.js]: I tried to use this setup but it looks like it is not compatible with the environment we use:

```javascript
TypeError: Cannot convert undefined or null to object
        at Function.getOwnPropertyDescriptors (<anonymous>)

      window.location = Object.defineProperties(
```

What do you think if we copy these:

```javascript
window.JP_CONNECTION_INITIAL_STATE = {
	userConnectionData: {
		currentUser: {
			wpcomUser: { Id: 99999, login: 'bobsacramento', display_name: 'Bob Sacrmaneto' },
		},
	},
};
```

Into our `global.window.JP_CONNECTION_INITIAL_STATE` mock?

File: jest.config.js (hunk 6 above)
Comment (geriux) [reply jest.config.js]: Updated in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5544/commits/9278461c173409d342520b8d7884c7f27b8f04cd and https://github.com/wordpress-mobile/gutenberg-mobile/pull/5544/commits/be3b94f1f4f716e706cee3dbbedb57d9647b5335

File: jest_setup.js (hunk 2 above)
Comment (geriux) [reply jest_setup.js]: Something like this https://github.com/wordpress-mobile/gutenberg-mobile/pull/5544/commits/29521810d3368f3f0a6b45f07a66cf8d42d78412

Comment (fluiddot) [reply jest_setup.js]: Yeah, that approach looks good to me, thanks! In the future, we'd need to revisit how to manage the Jetpack connection in GB-mobile and populate with data coming form the host (i.e. the app).

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 ! Thanks @geriux for fixing this 🙇 !

I tested locally by running `npm ci && npm run test` and all tests passed ✅ .

**Pull request merged**